include/zephyr/drivers/tee.h

```diff
@@ -0,0 +1,726 @@
+ *
+ * @param dev TEE device
+ * @param addr Address of the shared memory
+ * @param size Size of the shared memory region
+ * @param flags Flags to set registering parameters
+ * @param shm Return shared memory structure
+ *
+ * @retval -ENOSYS If callback was not implemented
+ *
+ * @retval 0       On success, negative on error
+ */
+__syscall int tee_shm_register(const struct device *dev, void *addr, size_t size,
+			       uint32_t flags, struct tee_shm **shm);
+
+static inline int z_impl_tee_shm_register(const struct device *dev, void *addr, size_t size,
+					  uint32_t flags, struct tee_shm **shm)
+{
+	flags &= ~TEE_SHM_ALLOC;
+	return tee_add_shm(dev, addr, 0, size, flags | TEE_SHM_REGISTER, shm);
+}
+
+/**
+ * @brief Unregister shared memory for Trusted Environment
+ *
+ * Unregisters shared memory for TEE
+ *
+ * @param dev TEE device
+ * @param shm Shared memory structure
+ *
+ * @retval -ENOSYS If callback was not implemented
+ *
+ * @retval 0       On success, negative on error
+ */
+__syscall int tee_shm_unregister(const struct device *dev, struct tee_shm *shm);
+
+static inline int z_impl_tee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+{
+	return tee_rm_shm(dev, shm);
+}
+
+/**
+ * @brief Allocate shared memory region for Trusted Environment
+ *
+ * Allocate shared memory for TEE
+ *
+ * @param dev TEE device
+ * @param size Region size
+ * @param flags to allocate region
+ * @param shm Return shared memory structure
+ *
+ * @retval -ENOSYS If callback was not implemented
+ *
+ * @retval 0       On success, negative on error
+ */
+__syscall int tee_shm_alloc(const struct device *dev, size_t size, uint32_t flags,
+			    struct tee_shm **shm);
+
+static inline int z_impl_tee_shm_alloc(const struct device *dev, size_t size, uint32_t flags,
+				       struct tee_shm **shm)
+{
+	return tee_add_shm(dev, NULL, 0, size, flags | TEE_SHM_ALLOC | TEE_SHM_REGISTER, shm);
+}
+
+/**
+ * @brief Free shared memory region for Trusted Environment
+ *
+ * Frees shared memory for TEE
+ *
+ * @param dev TEE device
+ * @param shm Shared memory structure
+ *
+ * @retval -ENOSYS If callback was not implemented
+ *
+ * @retval 0       On success, negative on error
+ */
+__syscall int tee_shm_free(const struct device *dev, struct tee_shm *shm);
+
+static inline int z_impl_tee_shm_free(const struct device *dev, struct tee_shm *shm)
+{
+	return tee_rm_shm(dev, shm);
+}
+
+/**
+ * @brief Receive a request for TEE Supplicant
+ *
+ * @param dev TEE device
+ * @param func Supplicant function
+ * @param num_params Number of parameters to be passed
+ * @param param List of the params for send/receive
+ *
+ * @retval -ENOSYS If callback was not implemented
+ *
+ * @retval 0       On success, negative on error
+ */
+__syscall int tee_suppl_recv(const struct device *dev, uint32_t func, unsigned int num_params,
+			     struct tee_param *param);
+
+static inline int z_impl_tee_suppl_recv(const struct device *dev, uint32_t func,
```

Comment:
> @lorc Supplicant polls driver via suppl_recv and wait blocked when no requests are available. When request is processed, it send response back via suppl_send, just like Linux tee_supplicant.

Ah, okay. Linux works in this way because it uses files abstraction. In case of Zehpyr, I believe, it is possible to return result right in `suppl_recv`, this might be easier than to support async logic in both OP-TEE driver and OP-TEE supplicant.